File: pipeline/schemas/sectors.py
"""Sectors/themes dataset contract (architecture §3.2 equity_structure and PRD themes module).

Filled by T03; this module only defines the contract.
"""

from __future__ import annotations

from pydantic import Field

from .envelope import BaseEnvelope, ContractModel, UTCDateTime


class SectorItem(ContractModel):
    """Sector or theme entry.

    No ``label``/``label_zh``: display labels live in ``src/i18n/locales/{en,zh-CN}/themes.json``,
    keyed by ``key`` (C-1/#102). The payload carries the key and the numbers; the frontend
    renders ``t(themes.<key>)`` and ``check:i18n`` catches a key with no Chinese label.

    ``constituents`` (themes only, #93) lists the theme's member symbols from
    ``config/themes.yaml`` — published so the Themes page can render them without a second
    data source. ``percentile_1y`` is the theme series' trailing-20-session return ranked in
    its trailing-252-session window (#86 §4); ``None`` with ``percentile_1y_obs`` below the
    configured minimum is "warming up", not "missing".
    """

    key: str = Field(min_length=1)
    change_1d: float | None = None
    change_1w: float | None = None
    change_1m: float | None = None
    percentile_1y: float | None = Field(default=None, ge=0.0, le=100.0, description="percentile within the ~1y history window (#70/#93)")
    percentile_1y_obs: int = Field(default=0, ge=0, description="observations behind the percentile (#70)")
    constituents: list[str] = Field(default_factory=list, description="theme member symbols (#93)")
    updated_at: UTCDateTime | None = None


class MemoryProxy(ContractModel):
    """Memory cycle proxy (review P0-1: MVP uses Micron/Hynix/Samsung share prices as a memory cycle proxy)."""

    label: str = Field(min_length=1)
    label_zh: str | None = None
    change_1w: float | None = None
    change_1m: float | None = None
    note: str | None = None
    updated_at: UTCDateTime | None = None


class SectorsDataset(ContractModel):
    """sectors.json payload."""

    sectors: list[SectorItem] = Field(default_factory=list)
    themes: list[SectorItem] = Field(default_factory=list)
    memory: MemoryProxy | None = None


class SectorsEnvelope(BaseEnvelope):
    payload: SectorsDataset

File: src/schemas/sectors.ts
import { z } from "zod";
import { datasetEnvelope, utcDateTime } from "./envelope";

export const SectorItem = z
  .object({
    key: z.string().min(1),
    label: z.string().min(1),
    label_zh: z.string().nullable(),
    change_1d: z.number().finite().nullable(),
    change_1w: z.number().finite().nullable(),
    change_1m: z.number().finite().nullable(),
    percentile_1y: z.number().finite().min(0).max(100).nullable(),
    percentile_1y_obs: z.number().int().min(0),
    updated_at: utcDateTime.nullable(),
  })
  .strict();

export const MemoryProxy = z
  .object({
    label: z.string().min(1),
    label_zh: z.string().nullable(),
    change_1w: z.number().finite().nullable(),
    change_1m: z.number().finite().nullable(),
    note: z.string().nullable(),
    updated_at: utcDateTime.nullable(),
  })
  .strict();

export const SectorsDataset = z
  .object({
    sectors: z.array(SectorItem),
    themes: z.array(SectorItem),
    memory: MemoryProxy.nullable(),
  })
  .strict();

export const SectorsEnvelope = datasetEnvelope(SectorsDataset);

export type SectorItem = z.infer<typeof SectorItem>;
export type MemoryProxy = z.infer<typeof MemoryProxy>;
export type SectorsDataset = z.infer<typeof SectorsDataset>;
export type SectorsEnvelope = z.infer<typeof SectorsEnvelope>;
